src/hashset.h

```diff
@@ -0,0 +1,188 @@
+     * when freeing. */
+    void (*trackMemUsage)(hashset *s, ssize_t delta);
+    /* Allow a hashset to carry extra caller-defined metadata. The extra memory
+     * is initialized to 0. */
+    size_t (*getMetadataSize)(void);
+    /* Flag to disable incremental rehashing */
+    unsigned instant_rehashing : 1;
+} hashsetType;
+
+typedef enum {
+    HASHSET_RESIZE_ALLOW = 0,
+    HASHSET_RESIZE_AVOID,
+    HASHSET_RESIZE_FORBID,
+} hashsetResizePolicy;
+
+typedef void (*hashsetScanFunction)(void *privdata, void *element);
+
+/* Constants */
+#define HASHSET_BUCKET_SIZE 64 /* bytes, the most common cache line size */
+
+/* Scan flags */
+#define HASHSET_SCAN_EMIT_REF (1 << 0)
+
+/* Iterator. To be able to use a stack-allocated iterator, it's not opaque. The
+ * fields in this struct should not be accessed though. Just use the hashset
+ * iterator functions. */
+typedef struct {
+    hashset *hashset;
+    hashsetBucket *bucket;
+    long index;
+    int table;
+    int pos_in_bucket;
+    uint64_t safe : 1;
+    union {
+        /* Unsafe iterator fingerprint for misuse detection. */
+        uint64_t fingerprint : 63;
+        /* Safe iterator temporary storage for bucket chain compaction. */
+        uint64_t last_seen_size : 63;
+    };
+} hashsetIterator;
+
+/* Position, used by some hashset functions such as two-phase insert and delete.
+ * This is not opaque, to be able to place it on the stack, but don't access the
+ * fields. Only use it with the hashset functions. */
+typedef struct {
+    void *bucket;
+    uint16_t pos_in_bucket;
+    uint16_t table_index;
+} hashsetPosition;
+
+/* --- Prototypes --- */
+
+/* Hash function (global seed) */
+void hashsetSetHashFunctionSeed(const uint8_t *seed);
+uint8_t *hashsetGetHashFunctionSeed(void);
+uint64_t hashsetGenHashFunction(const char *buf, size_t len);
+uint64_t hashsetGenCaseHashFunction(const char *buf, size_t len);
+
+/* Global resize policy */
+void hashsetSetResizePolicy(hashsetResizePolicy policy);
+
+/* Hashset instance */
+hashset *hashsetCreate(hashsetType *type);
+void hashsetRelease(hashset *s);
+void hashsetEmpty(hashset *s, void(callback)(hashset *));
+hashsetType *hashsetGetType(hashset *s);
+void *hashsetMetadata(hashset *s);
+size_t hashsetSize(hashset *s);
+size_t hashsetBuckets(hashset *s);
+size_t hashsetChainedBuckets(hashset *s, int table);
+size_t hashsetMemUsage(hashset *s);
+void hashsetPauseAutoShrink(hashset *s);
+void hashsetResumeAutoShrink(hashset *s);
+int hashsetIsRehashing(hashset *s);
+int hashsetIsRehashingPaused(hashset *s);
+void hashsetRehashingInfo(hashset *s, size_t *from_size, size_t *to_size);
+int hashsetRehashMicroseconds(hashset *s, uint64_t us);
+int hashsetExpand(hashset *s, size_t size);
+int hashsetTryExpand(hashset *s, size_t size);
+int hashsetExpandIfNeeded(hashset *s);
+int hashsetShrinkIfNeeded(hashset *s);
+hashset *hashsetDefragTables(hashset *s, void *(*defragfn)(void *));
```

Comment:
`hashsetDefragTables` handles only the allocation of the hashset struct and the tables. There are two tables only during rehashing and then one of the tables are new, so I imagine only one of them needs to be reallocated by defrag. It's not much to break into pieces here. Also: I modeled this after dict to avoid too many structural changes.